### PR TITLE
Internal/price in bank cache

### PR DESCRIPTION
**Pull request by IMEF-FEMI**
PR for the issue #417 

Authored by @sqhell 

cc: @jgur-psyops

### CODE REVIEW

File: programs/marginfi/src/instructions/kamino/withdraw.rs [hunk 1]
```diff
@@ -181,27 +181,37 @@ pub fn kamino_withdraw<'info>(
             amount: collateral_amount,
             close_balance: withdraw_all,
         });
+    }
```
Comment (IliaZyrin) [inline programs/marginfi/src/instructions/kamino/withdraw.rs]: Nitpick but just wanted to ask why this change? This way we load the marginfi account twice.

File: type-crate/src/types/bank_cache.rs [hunk 2]
```diff
@@ -35,7 +35,24 @@ pub struct BankCache {
...
+    /// * Only updated when instruction modifies shares AND uses oracle price
+    /// * Not updated for operations that don't require prices (e.g., deposit, repay)
+    /// * Price in USD (or quote currency) per token, as I80F48 (includes any bias applied)
+    /// * Zero if never updated
+    pub last_oracle_price: WrappedI80F48,
+
+    /// Unix timestamp (seconds) when last_oracle_price was last updated
+    /// * Used to determine staleness of cached price
+    /// * Zero if never updated
+    pub last_oracle_price_timestamp: i64,
```
Comment (IliaZyrin) [inline type-crate/src/types/bank_cache.rs:48]: I think we can just use existing Bank's `last_update` field instead of introducing the new one.
But we should then set it in Kamino withdraw (and maybe deposit - for consistency), because it's not updated there at the moment. It's only updated in `accrue_interest()` afaik, which is not called from Kamino instructions.

File: programs/marginfi/src/instructions/kamino/withdraw.rs [hunk 3]
```diff
@@ -181,27 +181,37 @@ pub fn kamino_withdraw<'info>(
...
-            );
-            risk_result?;
-            health_cache.program_version = PROGRAM_VERSION;
-
-            health_cache.set_engine_ok(true);
-            marginfi_account.health_cache = health_cache;
+    marginfi_account.lending_account.sort_balances();
+
+    // Note: during liquidating, we skip all health checks until the end of the transaction.
+    if !marginfi_account.get_flag(ACCOUNT_IN_RECEIVERSHIP) {
```
Comment (IliaZyrin) [inline programs/marginfi/src/instructions/kamino/withdraw.rs]: I think we should update the price cache in the "receivership" cases (liquidation/deleveraging) as well. It won't work for `withdraw_all` for now, but at least for the regular - yes.

On a side note, when [this](https://github.com/mrgnlabs/marginfi-v2/pull/446/files#diff-b0e8565f180e5eb656433fca060a9af25a6ea7406c5d3bd99248948387106be5L107) is done, we may also re-use the same price, for `withdrawn_equity` calculations and price cache updates, to not fetch it twice.

File: programs/marginfi/src/instructions/marginfi_account/liquidate.rs [hunk 4]
```diff
@@ -379,10 +379,6 @@ pub fn lending_account_liquidate<'info>(
                 .ok_or(MarginfiError::MathError)?
                 .into();
 
-        asset_bank.update_bank_cache(group)?;
```
Comment (IliaZyrin) [inline programs/marginfi/src/instructions/marginfi_account/liquidate.rs:382]: Not a big deal but I would not move these and instead used the `risk_engine` from [here](https://github.com/mrgnlabs/marginfi-v2/pull/446/files#diff-0784a7c2bddc126f67aa6e98cb8f72639d8e74289f0c7725c5d0288eba045ccdR165).

E.g. replace that part with:
```
let risk_engine = RiskEngine::new(&liquidatee_marginfi_account, liquidatee_remaining_accounts)?;
    let (pre_liquidation_health, _, _) = risk_engine
            .check_pre_liquidation_condition_and_get_account_health(
                Some(&ctx.accounts.liab_bank.key()),
                &mut None,
                false,
            )?;
            
```

File: programs/marginfi/src/instructions/marginfi_group/handle_bankruptcy.rs [hunk 5]
```diff
@@ -175,8 +182,6 @@ pub fn lending_pool_handle_bankruptcy<'info>(
     )?
     .repay(bad_debt)?;
 
-    bank.update_bank_cache(group)?;
```
Comment (IliaZyrin) [inline programs/marginfi/src/instructions/marginfi_group/handle_bankruptcy.rs:178]: This must not be moved from here because it must be called after all manipulations with the bank such as interest accrual and loss socialization.

File: programs/marginfi/src/state/marginfi_account.rs [hunk 6]
```diff
@@ -604,6 +606,29 @@ impl<'info> RiskEngine<'_, 'info> {
...
+
+        let bank = bank_account.bank.load()?;
+        let (price_feed_res, _) = bank_account.try_get_price_feed();
+        let price_feed = price_feed_res?;
+
+        let price = price_feed.get_price_and_confidence_of_type(
+            OraclePriceType::RealTime,
+            bank.config.oracle_max_confidence,
+        )?;
+        check!(price.price > I80F48::ZERO, MarginfiError::ZeroAssetPrice);
```
Comment (IliaZyrin) [inline programs/marginfi/src/state/marginfi_account.rs]: Not sure we should check for zero here. In theory nothing should prevent you from withdrawing your zero-priced assets from the bank.

File: programs/marginfi/src/state/price.rs [hunk 7]
```diff
@@ -634,21 +656,36 @@ impl PriceAdapter for SwitchboardPullPriceFeed {
         let price = self.get_price()?;
 
         match bias {
+            None => Ok(price),
```
Comment (IliaZyrin) [inline programs/marginfi/src/state/price.rs]: Nitpick (at least for the future) but I would not change such things because it bloats the PR and takes reviewers' time to check it and only see that nothing has changed except for cosmetics.

File: programs/marginfi/src/utils/general.rs [hunk 8]
```diff
@@ -363,6 +366,70 @@ pub fn fetch_asset_price_for_bank<'info>(
     Ok(price)
 }
 
+/// Fetch liability price for a given bank from a properly structured remaining accounts slice.
+/// Uses High bias for conservative liability valuation. Errors if the bank is not found or the price is zero.
+pub fn fetch_liability_price_for_bank<'info>(
```
Comment (IliaZyrin) [inline programs/marginfi/src/utils/general.rs]: It's not used anywhere, right? Let's remove it then.

File: type-crate/src/types/bank_cache.rs (hunk 2 above)
Comment (IMEF-FEMI) [reply type-crate/src/types/bank_cache.rs:48]: we might have a challenge with this as the bank cache updates (for rates) and the price updates usually does not necessarily happen at the same time. as sometimes we update the rates without updating the cached price... like in deposit function

Comment (IliaZyrin) [reply type-crate/src/types/bank_cache.rs:48]: True, makes sense.

File: programs/marginfi/src/instructions/kamino/withdraw.rs [hunk 9]
```diff
@@ -96,8 +97,12 @@ pub fn kamino_withdraw<'info>(
 
         let in_receivership = marginfi_account.get_flag(ACCOUNT_IN_RECEIVERSHIP);
         let price = if in_receivership {
-            let price =
-                fetch_asset_price_for_bank(&bank_key, &bank, &clock, ctx.remaining_accounts)?;
+            let price = fetch_asset_price_for_bank_low_bias(
```
Comment (IliaZyrin) [inline programs/marginfi/src/instructions/kamino/withdraw.rs:100]: I think ideally we want a non-biased price here. It's not critical at all but if it also helps to reduce the number of `fetch_price_*` functions, let's do it.

File: programs/marginfi/src/state/bank.rs [hunk 10]
```diff
@@ -560,6 +568,30 @@ impl BankImpl for Bank {
...
+    fn update_cache_price(
+        &mut self,
+        oracle_price: Option<OraclePriceWithConfidence>,
+    ) -> MarginfiResult<()> {
+        if let Some(price_with_confidence) = oracle_price {
+            self.cache.last_oracle_price = price_with_confidence.price.into();
+            self.cache.last_oracle_price_confidence = price_with_confidence.confidence.into();
+            self.cache.last_oracle_price_timestamp = Clock::get()?.unix_timestamp;
+        } else {
+            msg!("price cache update failed");
```
Comment (IliaZyrin) [inline programs/marginfi/src/state/bank.rs]: I think this message is confusing given that [there are cases](https://github.com/mrgnlabs/marginfi-v2/pull/446/files#diff-823904705faab762f8a1fe01b204e4161589a8e3bc6bc24a6a85d4e4e33b45deR219) where we explicitly expect the price to be `None`. So there is nothing to _fail_ here, it's an expected behavior.

On the other hand, in Kamino Withdraw we simply [do not call](https://github.com/mrgnlabs/marginfi-v2/pull/446/files#diff-b0e8565f180e5eb656433fca060a9af25a6ea7406c5d3bd99248948387106be5R210) this if the price is `None`. If we apply the same approach to all the other "maybe" cases, then we can leave the message as it is.

File: programs/marginfi/tests/user_actions/bank_cache.rs [hunk 11]
```diff
@@ -0,0 +1,312 @@
...
+        sol_bank_post.last_update >= sol_bank_pre.last_update,
+        "Bank's last_update should be updated after bankruptcy"
+    );
+
+    Ok(())
+}
+
+/// Test that bank's last_update field is set when bank is modified
+/// This verifies that last_update can be used to determine price staleness
+/// (since we removed the dedicated last_oracle_price_timestamp field)
```
Comment (IliaZyrin) [inline programs/marginfi/tests/user_actions/bank_cache.rs:304]: You didn't remove `last_oracle_price_timestamp` though: https://github.com/mrgnlabs/marginfi-v2/pull/446/files#diff-6abc5d6d44fd41e73ed6d1d7c6318c6442f7122712cc1ea4c29c791335ee3b16R48

File: tests/utils/user-instructions.ts [hunk 12]
```diff
@@ -572,6 +578,29 @@ export const healthPulse = (
     .instruction();
 };
 
+/**
+ * Creates a bank price pulse instruction. This ix refreshes the cached oracle price
+ * for a given bank without modifying user positions.
+ * * `remaining` - pass the oracle accounts required for this bank
+ *   in the same order as used for other oracle reads for that bank.
+ */
+export const bankPricePulse = (
```
Comment (IliaZyrin) [inline tests/utils/user-instructions.ts]: Nitpick but I would name it `pulseBankPrice` for consistency here.

File: tests/z02_lastOraclePrice.spec.ts [hunk 13]
```diff
@@ -0,0 +1,250 @@
...
+
+    const bankAfter = await bankrunProgram.account.bank.fetch(banks[1]);
+    const afterFields = readCacheFields(bankAfter.cache);
+
+    // update_bank_cache was called with None, so price/confidence should be unchanged
+    assertI80F48Equal(afterFields.price, beforeFields.price);
+    assertI80F48Equal(afterFields.conf, beforeFields.conf);
+  });
+
+  it("(user 0) full repay resets bank cache when liabilities go to zero", async () => {
```
Comment (IliaZyrin) [inline tests/z02_lastOraclePrice.spec.ts]: Just curious what was the reason for this logic?

File: programs/marginfi/src/instructions/kamino/withdraw.rs (hunk 9 above)
Comment (jgur-psyops) [reply programs/marginfi/src/instructions/kamino/withdraw.rs:100]: I think we did low here in case price > 0 but price - confidence  = 0. Not sure if that's actually possible tho. We need this "low bias" one for liquidation (it's actually the original, I think, and the others were written later).

File: programs/marginfi/src/state/bank.rs (hunk 10 above)
Comment (jgur-psyops) [reply programs/marginfi/src/state/bank.rs]: Fair point, I'll just silence the message, it's a waste of compute anyways.

File: programs/marginfi/tests/user_actions/bank_cache.rs (hunk 11 above)
Comment (jgur-psyops) [reply programs/marginfi/tests/user_actions/bank_cache.rs:304]: I think let's just leave it. There are these weird edge cases where the rest of the cache updates and the price cache doesn't, so this does give you some extra utility.

File: tests/z02_lastOraclePrice.spec.ts (hunk 13 above)
Comment (jgur-psyops) [reply tests/z02_lastOraclePrice.spec.ts]: Good point, maybe we shouldn't. I think when we originally wrote `update_bank_cache`

```
        if (total_assets_amount == I80F48::ZERO) || (total_liabilities_amount == I80F48::ZERO) {
            self.cache = BankCache::default();
            return Ok(());
        }
```

this reset to zero made sense. Now that there's other stuff in the cache what we really want to do is just reset the rates, not the price cache. I'll make it do that.

File: programs/marginfi/src/instructions/kamino/withdraw.rs (hunk 9 above)
Comment (IliaZyrin) [reply programs/marginfi/src/instructions/kamino/withdraw.rs:100]: I don't think it's possible, but even if so, we still want that "> 0" price. And yeah, we probably just reused the function which was already available, makes sense.
